src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -88,8 +96,18 @@ classTrack_processUnloads(JNIEnv *env)
         return NULL;
     }
     struct bag* deleted = deletedSignatures;
-    deletedSignatures = bagCreateBag(sizeof(char*), 10);
+    deletedSignatures = NULL;
+    debugMonitorExit(classTrackLock);
```

Comment:
This looks risky as the critical section is broken and the NULL deleted signatures exposed. If `cbTrackingObjectFree` occurs while this is true then you will lose the record of the deleted signature.

Alternatively you could allow for lock-free reading of `deletedSignatures`, preemptively allocate a new bad if needed then take the lock. Or even use the lock to read `deletedSignatures` to determine if a new bag is needed, then drop the lock, create the bag, take the lock and re-check everything.

Reply:
Agreed.

Reply:
Fixed